src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs

```diff
@@ -61,7 +61,21 @@ internal SocketConnection(Socket socket, MemoryPool<byte> memoryPool, PipeSchedu
         public override MemoryPool<byte> MemoryPool { get; }
         public override PipeScheduler InputWriterScheduler => _scheduler;
         public override PipeScheduler OutputReaderScheduler => _scheduler;
-        public override long TotalBytesWritten => Interlocked.Read(ref _totalBytesWritten);
+
+        public override long TotalBytesWritten
+        {
+            get
+            {
+                if (IntPtr.Size == sizeof(long))
+                {
+                    return _totalBytesWritten;
```

Comment:
Shouldn't Interlocked.Read already optimized to this on 64-bit systems? @benaadams

Reply:
Nah, [`Interlocked.Read`](https://github.com/dotnet/coreclr/blob/d73a35065ec05a8179c8b812986bd9708001782c/src/System.Private.CoreLib/src/System/Threading/Interlocked.cs#L157-L160) always does what it says:
```csharp
public static partial class Interlocked
{
    public static long Read(ref long location)
    {
        return Interlocked.CompareExchange(ref location, 0, 0);
    }
}
```
You are thinking of [`Volatile.Read`](https://github.com/dotnet/coreclr/blob/d73a35065ec05a8179c8b812986bd9708001782c/src/System.Private.CoreLib/src/System/Threading/Volatile.cs#L110-L160)
```csharp
public static partial class Volatile
{
#if BIT64
    public static long Read(ref long location)
    {
        // The VM will replace this with a more efficient implementation.
        var value = location;
        Interlocked.MemoryBarrier();
        return value;
    }
#else
    public static long Read(ref long location)
    {
        // On 32-bit machines, we use this implementation, since an ordinary volatile read
        // would not be atomic.
        //
        // On 64-bit machines, the VM will replace this with a more efficient implementation.
        return Interlocked.CompareExchange(ref location, 0, 0);
    }
#endif
}
```

Reply:
I probably should have just looked for myself. Thanks though 😄

Reply:
Though it should probably be just changed to `Volatile.Read` rather than using `Interlocked.Read`

Reply:
Will that prevent tearing?

Reply:
^--- has a `#if BIT64` switch that uses `Interlocked.CompareExchange` in the 32bit path

Reply:
For reference the following code produces 

```c#
public class Bench
{
    private long _value = 42;

    [Benchmark(Baseline = true)]
    public long WithInterlocked() => Interlocked.Read(ref _value);

    [Benchmark]
    public long DirectReadOnx64()
    {
        if (IntPtr.Size == sizeof(long))
        {
            return _value;
        }
        else
        {
            return Interlocked.Read(ref _value);
        }
    }

    [Benchmark]
    public long VolatileRead() => Volatile.Read(ref _value);
}
```

this dasm

```asm
; WithInterlocked
00007FFC0E951510  cmp         dword ptr [rcx],ecx
00007FFC0E951512  lea         rdx,[rcx+8]
00007FFC0E951516  xor         ecx,ecx
00007FFC0E951518  xor         eax,eax
00007FFC0E95151A  lock cmpxchg qword ptr [rdx],rcx
00007FFC0E95151F  ret
;------------------------------------------------------------------------------
; DirectReadOnx64
00007FFC0E951530  mov         rax,qword ptr [rcx+8]
00007FFC0E951534  ret
;------------------------------------------------------------------------------
; VolatileRead
00007FFC0E951550  lea         rax,[rcx+8]
00007FFC0E951554  mov         rax,qword ptr [rax]
00007FFC0E951557  ret
```

So `DirectReadOnx64` and `VolatileRead` is "the same" code.

For bench-results I get this:

|          Method |      Mean |     Error |    StdDev | Scaled |
|---------------- |----------:|----------:|----------:|-------:|
| WithInterlocked | 5.9178 ns | 0.1028 ns | 0.0912 ns |  1.000 |
| DirectReadOnx64 | 0.0260 ns | 0.0229 ns | 0.0215 ns |  0.004 |
|    VolatileRead | 0.0194 ns | 0.0191 ns | 0.0179 ns |  0.003 |

So quite as expected.

I'll update to `Volatile.Read`.